### PR TITLE
Add thrift definition for low cardinality string global dict

### PR DESCRIPTION
**Pull request by kangkaisen**
For https://github.com/StarRocks/starrocks/issues/559, Add some thrift definitions

### CODE REVIEW

File: gensrc/thrift/Data.thrift [hunk 1]
```diff
@@ -92,5 +92,9 @@ struct TStatisticData {
     9: optional string max
     10: optional string min
     11: optional string histogram
+    // global dict for low cardinality string
+    12: optional binary dict
+    // the latest tablet version for this table
+    13: optional i64 meta_version
```
Comment (Seaven) [inline gensrc/thrift/Data.thrift:98]: Why not use statistic_version？

File: gensrc/thrift/Data.thrift [hunk 2]
```diff
@@ -92,5 +92,9 @@ struct TStatisticData {
     9: optional string max
     10: optional string min
     11: optional string histogram
+    // global dict for low cardinality string
+    12: optional binary dict
```
Comment (imay) [inline gensrc/thrift/Data.thrift:96]: Why not define a struct to describe this dict, in order to make it readable.

File: gensrc/thrift/PlanNodes.thrift [hunk 3]
```diff
@@ -745,6 +738,15 @@ struct TProjectNode {
     2: optional map<Types.TSlotId, Exprs.TExpr> common_slot_map
 }
 
+struct TMetaScanNode {
+    // column id to column name
+    1: optional map<i32, string> id_to_names
+}
+
+struct TDecodeNode {
+    1: optional map<i32, i32> dict_id_to_string_ids
```
Comment (imay) [inline gensrc/thrift/PlanNodes.thrift:747]: Add comment for these two ids?

File: gensrc/thrift/Planner.thrift [hunk 4]
```diff
@@ -65,6 +65,13 @@ struct TPlanFragment {
   // sink) in a single instance of this fragment. This is used for an optimization in
   // InitialReservation. Measured in bytes. required in V1
   8: optional i64 initial_reservation_total_claims
+  20: optional list<TGlobalDict> global_dicts
+}
+
+struct TGlobalDict {
+    1: optional i32 columnId
+    2: optional list<string> strings
```
Comment (imay) [inline gensrc/thrift/Planner.thrift:73]: binary is better?

File: gensrc/thrift/Planner.thrift [hunk 5]
```diff
@@ -65,6 +65,13 @@ struct TPlanFragment {
...
+    3: optional list<i32> ids
```
Comment (DeepThinker666) [inline gensrc/thrift/Planner.thrift:74]: what is ids for?

File: gensrc/thrift/Planner.thrift (hunk 4 above)
Comment (kangkaisen) [reply gensrc/thrift/Planner.thrift:73]: Why binary is better.

File: gensrc/thrift/Planner.thrift (hunk 5 above)
Comment (kangkaisen) [reply gensrc/thrift/Planner.thrift:74]: The dict id.

File: gensrc/thrift/Data.thrift (hunk 1 above)
Comment (starrocks-xueli) [inline gensrc/thrift/Data.thrift:98]: maybe it's better to use `tablet_version`？

File: gensrc/thrift/Planner.thrift (hunk 4 above)
Comment (imay) [reply gensrc/thrift/Planner.thrift:73]: String's content is character, binary's is byte.
What we needs here is binary. For example, we can use binary to store int64.

File: gensrc/thrift/Data.thrift (hunk 2 above)
Comment (imay) [reply gensrc/thrift/Data.thrift:96]: What's the binary content?

Comment (kangkaisen) [reply gensrc/thrift/Data.thrift:96]:
```
agg function: dict_merge
input_type: TYPE_ARRAY (VARCHAR)
return_type: TYPE_VARCHAR (binary)

|dict-code: 4 bytes|string-size 4 bytes|string-content|...
```

**Review by laotan332 — DISMISSED** [review]

File: gensrc/thrift/Planner.thrift (hunk 4 above)
Comment (kangkaisen) [reply gensrc/thrift/Planner.thrift:73]: We use `HashMap<String, Integer>` in FE and BE, use string, we handle it easily in FE and BE both. if use binary, the will have Little-Endian and Big-Endian issue. The map is small, so I think there is no performance issue.

File: gensrc/thrift/Data.thrift (hunk 1 above)
Comment (kangkaisen) [reply gensrc/thrift/Data.thrift:98]: In fact, we get the version from the max partition VisibleVersion. Which is equal the the latest tablet version, the latest partition version, the latest table version. So we use the general conception: meta version.
```
            long version = table.getPartitions().stream().map(Partition::getVisibleVersion)
                    .max(Long::compareTo).orElse(0L);
```

File: gensrc/thrift/Planner.thrift (hunk 4 above)
Comment (imay) [reply gensrc/thrift/Planner.thrift:73]: It's not about performance. Rather, it is about expressiveness.Our dictionary encoding can be used not just for strings, but for other types. For example, in the future we will support dictionary encoding of int64_t.
And with regard to the big endian or little endian, it is not a question of whether to use string or binary.

File: gensrc/thrift/Data.thrift (hunk 2 above)
Comment (imay) [reply gensrc/thrift/Data.thrift:96]: Why not define the content in thrift format? For performance? It seems not a performance issue.

Comment (kangkaisen) [reply gensrc/thrift/Data.thrift:96]: The dict content is the aggregate function result. Why could be thrift format ???

Comment (imay) [reply gensrc/thrift/Data.thrift:96]: There are several benefits to using thrift.
1. clearer definitions and more intuitive for the reader.
2. better compatibility. In the future when the structure changes, it is easier to be backwards and forwards compatible.
3. Better programming interface. No additional serialisation and deserialisation procedures are needed, and the structure can be modified directly.
4. More cross-language friendly. It is easier to manipulate structures when they need to be manipulated across languages.

If nothing else is bad, I would prefer to define it as a thrift structure for future extensibility.

**Review by stdpain — DISMISSED** [review]

File: gensrc/thrift/Planner.thrift [hunk 6]
```diff
@@ -65,6 +66,13 @@ struct TPlanFragment {
   // sink) in a single instance of this fragment. This is used for an optimization in
   // InitialReservation. Measured in bytes. required in V1
   8: optional i64 initial_reservation_total_claims
+  20: optional list<TGlobalDict> global_dicts
+}
+
+struct TGlobalDict {
```
Comment (stdpain) [inline gensrc/thrift/Planner.thrift:72]: we defined it in Data.thrift, we could remove it

Comment (kangkaisen) [reply gensrc/thrift/Planner.thrift]: Yes. I thought I removed it...

**Review by imay — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**